File: src/main/java/org/shark/example/config/event/AsyncTaskDecorator.java
package org.shark.example.config.event;

import org.slf4j.MDC;
import org.springframework.core.task.TaskDecorator;

import java.util.Map;

public class AsyncTaskDecorator implements TaskDecorator {
    @Override
    public Runnable decorate(Runnable runnable) {
        Map<String,String> map = MDC.getCopyOfContextMap();
        return () -> {
            try{
                MDC.setContextMap(map);
                runnable.run();
            } finally {
                MDC.clear();
            }
        };
    }
}

File: src/main/kotlin/org/shark/example/config/event/AsyncTaskDecorator.kt
package org.shark.example.config.event

import org.slf4j.MDC
import org.springframework.core.task.TaskDecorator

class AsyncTaskDecorator : TaskDecorator {
    override fun decorate(runnable: Runnable): Runnable {
        val map = MDC.getCopyOfContextMap()
        return Runnable {
            try {
                if (map != null) {
                    MDC.setContextMap(map)
                }
                runnable.run()
            } finally {
                MDC.clear()
            }
        }
    }
}
